crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2391,75 +2434,100 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
-    fn infer_name_expression(&mut self, name: &ast::ExprName) -> Type<'db> {
-        let ast::ExprName { range: _, id, ctx } = name;
+    /// Infer the type of a [`ast::ExprName`] expression, assuming a load context.
+    fn infer_name_load(&mut self, name: &ast::ExprName) -> Type<'db> {
+        let ast::ExprName {
+            range: _,
+            id,
+            ctx: _,
+        } = name;
+
         let file_scope_id = self.scope().file_scope_id(self.db);
+        let use_def = self.index.use_def_map(file_scope_id);
+        let symbol = self
+            .index
+            .symbol_table(file_scope_id)
+            .symbol_id_by_name(id)
+            .expect("Expected the symbol table to create a symbol for every Name node");
+        // if we're inferring types of deferred expressions, always treat them as public symbols
+        let (definitions, may_be_unbound) = if self.is_deferred() {
+            (
+                use_def.public_bindings(symbol),
+                use_def.public_may_be_unbound(symbol),
+            )
+        } else {
+            let use_id = name.scoped_use_id(self.db, self.scope());
+            (
+                use_def.bindings_at_use(use_id),
+                use_def.use_may_be_unbound(use_id),
+            )
+        };
 
-        match ctx {
-            ExprContext::Load => {
-                let use_def = self.index.use_def_map(file_scope_id);
-                let symbol = self
-                    .index
-                    .symbol_table(file_scope_id)
-                    .symbol_id_by_name(id)
-                    .expect("Expected the symbol table to create a symbol for every Name node");
-                // if we're inferring types of deferred expressions, always treat them as public symbols
-                let (definitions, may_be_unbound) = if self.is_deferred() {
-                    (
-                        use_def.public_bindings(symbol),
-                        use_def.public_may_be_unbound(symbol),
-                    )
-                } else {
-                    let use_id = name.scoped_use_id(self.db, self.scope());
-                    (
-                        use_def.bindings_at_use(use_id),
-                        use_def.use_may_be_unbound(use_id),
-                    )
-                };
```

Comment:
The diff here is messy, but this is just: taking the `ExprContext::Load` branch and moving it into a dedicated method (`infer_name_load`), then calling it from the aforementioned branch.